### ISSUE TITLE
Use $container->call() for calling DI\factory functions

### ISSUE THREAD

**Issue by ewinslow** [PHP-DI/PHP-DI#197]
Right now when you use DI\factory you have to do:

```php
factory(function(Container $c) {
  return new Thing($c->get(Service::class));
});
```

What I'd really like to do is this:

```php
factory(function(Service $s) {
  return new Thing($s);
});
```

**Comment by mindplay-dk** [comment]
As far as performance, whatever needs to get resolved via argument type-hints is strictly stuff that would need to get resolved by explicit container lookups inside the definition anyhow - likely the performance difference is going to be marginal?

**Comment by mnapoli** [comment]
@mindplay-dk the thing is that it will use reflection on the closure to guess the parameters. Also `$container->call()` goes through the [Invoker](https://github.com/PHP-DI/Invoker) library and I haven't worked on performances on this one (can't guarantee it's the fastest thing ever).

Benchmarks would probably help to get an idea.

Also could you have a look at #239 and let me know if you think it's a good idea? Because it could be useful with this feature (for example for parameters/non-object dependencies).

**Comment by mindplay-dk** [comment]
Okay, our entire framework has been refactored now, so I can profile tomorrow at work for comparison with profiles I ran earlier this week - I have a feeling this is faster than even the lightweight container we had in there before, and the overhead was already negligible, but I will let you know. I can benchmark the total roundtrip time with/without this simple one-line change to see how that performs too. (I'm less interested in raw container performance and more interested in how it performs in real life - even if it was 20% slower but added < 1% total round-trip overhead, for example, I would care more about productivity than bare bones performance of a component that is already "cheap".)

**Comment by mnapoli** [comment]
Great! If you can, use the latest release (5.0.3) as I added a few performance improvements yesterday.

**Comment by mnapoli** [comment]
@mindplay-dk did you have a chance to do performance tests or profiling with this feature (i.e. use `Container::call()` for invoking factory definitions)?

**Comment by mindplay-dk** [comment]
Sorry, no - this may sound odd, after all my excitement about PHP-DI, but we ended up switching to a much simpler container library I had been working on myself, incorporating some of the missing ideas we got from using your library, and some of the absent features we've discussed earlier, including the Angular-style injections.

To be clear, it's not that there was anything wrong with your library - quite the contrary! We just wanted something simpler, less like a DSL, relying on heavy use of closures, primarily in order to get full IDE support, code inspections, safety for automated refactoring and such. Not that we couldn't have done that with PHP-DI - but for performance reasons, we had already turned off annotation support, and for other reasons, we had turned off auto-wiring, so in the end, we really weren't using most of PHP-DI at all.

We have a lot of juniors on our team, and we wanted something with a much smaller API and fewer concepts, in order to get them up to speed faster, and in order to avoid everyone using different features and patterns to do similar things, so we settled for something that supports only the patterns we want our team to use.

Thanks for your time and dedication to this project, you're doing excellent work! PHP-DI still gets my highest recommendations :-)

**Comment by mnapoli** [comment]
Thanks for letting me know ;) That's great to hear feedback like this, you might not be using this library but I'm always interested to learn about use cases and other solutions (and I understand your decision, it makes perfect sense). You mentioned angular-style injections (#239) is it nice to use (and read)? Is it limited/worth it?

**Comment by mindplay-dk** [comment]
> You mentioned angular-style injections (#239) is it nice to use (and read)? Is it limited/worth it?

It was, until it wasn't ;-)

I actually started out with a syntax like this:

```php
$container->call(function (Cache $cache, $db_name) {
    // ...
}, ["db.cache", "db.name"]);
```

But more often than not, I found that we needed to specify a mix of component name references and values, and I didn't want two different array arguments for that, so we tried this:

```php
$container->call(function (Cache $cache, $db_name) {
    // ...
}, [$container->value($cache), "db.name"]);
```

In other words, you had to box values to make them discernible to the container.

That seemed counter-intuitive, so we ended up doing the opposite:

```php
$container->call(function (Cache $cache, $db_name) {
    // ...
}, ["db.cache", $container->ref("db.name")]);
```

Or in other words, component references have to be boxed.

Documentation is [here](https://github.com/mindplay-dk/unbox) if you'd like to see what else we came up with.

**Comment by mnapoli** [comment]
Good to know then thanks :)

**Comment by Trainmaster** [comment]
Any news on this topic? I'd love to see things like 
```php
$container->set('Foo', \DI\factory([FooFactory::class, 'createInstance', ['param1', 'param2']]));
```
working.

**Comment by mnapoli** [comment]
@Trainmaster there is #333 and more recently #345 (which includes #333) regarding what's discussed in this issue. It is planned to merge them and release it as 5.2.0, I'm just lacking a bit of time at the moment to review everything and test performances as much as I would like to. Probably this week or the next!

**Comment by mnapoli** [comment]
I forgot to close this, this was implemented in 5.2.
